### PR TITLE
 Exclude aufs whiteout files from docker extraction (Alternative C/libarchive impl.)

### PR DESCRIPTION
**Pull request by dtrudg**
**Description of the Pull Request (PR):**

**This is an alternative implementation of PR #1106 using C code and libarchive to provide libexec/bin/docker-extract for aufs whiteout handling instead of complex shell code.**

Docker layer tars can contain .wh.* files/dirs that are whiteout files from aufs. When extracted to a singularity image this causes problems. E.g.

singularity pull docker://nvidia/digits:6.0
singularity run digits-6.0.img

... will fail with python errors related to permission issues as python tries to scan .wh* files and dirs.

The aufs whiteout files instruct container software how to remove files from previous layers, that were removed during a later Dockerfile 'RUN' step. These removals must be performed by Singularity to ensure a container contains the same content as when run with docker.

The issue is illustrated with the test container at this Git repository:

https://github.com/dctrud/docker-aufs-sanity

**This fixes or addresses the following GitHub issues:**

- Ref: #571 #884 #1132 (part of issue)

**Checkoff for all PRs:**

- [X] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [X] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [X] I have tested this PR locally with a `make test`
- [X] This PR is NOT against the project's master branch
- [X] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [X] This PR is ready for review and/or merge

Attn: @singularityware-admin

### CODE REVIEW

**Comment by dtrudg** [comment]
Travis failure on CentOS7 is odd. Don't think this is due to issue in this commit, since 03346fc passed

```
 + sudo singularity build --isolated --writable /tmp/stest.6UkvFY/container2 ../examples/busy    (retval=0) OK
 + sudo singularity build --isolated /tmp/stest.6UkvFY/container /tmp/stest.6UkvFY/Singularit    (retval=0) OK
 + singularity exec /tmp/stest.6UkvFY/container true                                             (retval=0) OK
 + singularity exec /tmp/stest.6UkvFY/container false                                            (retval=1) OK
 + singularity exec /tmp/stest.6UkvFY/container test -f /.singularity.d/runscript                (retval=0) OK
 + singularity exec /tmp/stest.6UkvFY/container test -f /.singularity.d/env/01-base.sh           (retval=0) OK
 + singularity exec /tmp/stest.6UkvFY/container test -f /.singularity.d/actions/shell            (retval=0) OK
 + singularity exec /tmp/stest.6UkvFY/container test -f /.singularity.d/actions/exec             (retval=0) OK
 + singularity exec /tmp/stest.6UkvFY/container test -f /.singularity.d/actions/run              (retval=0) OK
 + singularity exec /tmp/stest.6UkvFY/container test -L /environment                             (retval=0) OK
 + singularity exec /tmp/stest.6UkvFY/container test -L /singularity                             (retval=0) OK
 + sudo singularity -x build --isolated --force --sandbox /tmp/stest.6UkvFY/container2 ../exa    (retval=0) OK
 + sudo singularity build --isolated /tmp/stest.6UkvFY/container /tmp/stest.6UkvFY/Singularit  (retval=255) ERROR
Using container recipe deffile: /tmp/stest.6UkvFY/Singularity
Building into existing container: /root/build
Using container recipe deffile: Singularity
Sanitizing environment
[0mAdding base Singularity environment to container
Exporting contents of container2 to /root/build
Found an existing definition file
Adding a bootstrap_history directory
Finalizing Singularity container
Calculating final size for metadata...
Skipping checks
Singularity container built: /root/build
Cleaning up...
Building Singularity FS image...
Write failed because Bad file descriptor
FATAL ERROR:Failed to write to output filesystem
[91mERROR  : Image path doesn't exists
[0m[31mABORT  : Retval = 255
[0mCleaning up...
Full output in: /tmp/stest.6UkvFY
ERROR: Failed running test: 20-build.sh
make: *** [test] Error 1
```

File: src/docker-extract.c
Comment (GodloveD) [inline src/docker-extract.c]: You should probably get this from the registry rather than pulling it directly from the environment.  I don't know if that prevents a race condition or something, but at least for consistency's sake.

File: src/docker-extract.c
Comment (GodloveD) [inline src/docker-extract.c]: Seems like this check could probably be first?  If the input isn't right why bother doing anything else?

File: libexec/cli/build.exec
Comment (GodloveD) [inline libexec/cli/build.exec]: Are these echos debugging cruft?

File: src/docker-extract.c
Comment (GodloveD) [inline src/docker-extract.c]: nitpicky, but maybe name this variable `retval` instead for consistency.

File: src/docker-extract.c
Comment (GodloveD) [inline src/docker-extract.c:212]: Naive question.  Can we always be sure that the block size will be 10240?

File: src/docker-extract.c
Comment (GodloveD) [inline src/docker-extract.c]: Could the closes above be grouped so that only a single if statement is needed?

File: src/docker-extract.c
Comment (GodloveD) [inline src/docker-extract.c]: As a general rule I think we want to avoid declaring variables in the middle of scope like this.  Some of the C programmers on the team can give more insight on best practices.

File: .travis/rpmbuild_test
Comment (GodloveD) [inline .travis/rpmbuild_test]: I think you may need to remove the dep here and put it in the spec and deb files instead.  See https://github.com/singularityware/singularity/pull/1198 and https://github.com/singularityware/singularity/pull/1204 for examples of what we were doing wrong and how to fix.

File: src/docker-extract.c
Comment (GodloveD) [inline src/docker-extract.c:25]: Based on the logic here I was expecting directories marked with an opq file to be deleted.  But I see that those directories still exist in a DIGITS image created using this branch.  Moreover in most cases (but not all) the contents of directories marked with a opq file is identical to the contents of the directory if the image is built with the current dist version instead.  What am I missing here?  Is the directory recreated when the current layer is applied?  Is that the expected/desired behavior?

**Review by GodloveD — CHANGES_REQUESTED** [review]
Hi @dctrud.  Thanks again for taking the time to identify and provide a solution for this issue.  I've taken a first pass at review and I think I've identified some things that should be clarified and fixed.  Once we get through this round I'm going to ask one of our more experienced C devs to go over the code for best practices and security checking etc.  Thanks again.  I look forward to hearing more.  😄

File: src/docker-extract.c
Comment (GodloveD) [inline src/docker-extract.c]: Sorry, meant to include this in the previous review.  I'm going to advocate that we always use curly braces for if statements even if they are just simple one liners like this.  Otherwise my brain skips a beat when I'm reading and I have to stop and do a double take. lol

File: libexec/cli/build.exec
Comment (dtrudg) [reply libexec/cli/build.exec]: Ahh yes. Whoops.

File: src/docker-extract.c
Comment (dtrudg) [reply src/docker-extract.c:25]: An opaque directory is one that was removed, but then recreated later. It`s opaque in that the files from earlier layers don’t show through. Believe I’m implementing this correctly. Directory is remove when processing the white outs then will reappear when layer tar is extracted.

File: src/docker-extract.c
Comment (dtrudg) [reply src/docker-extract.c]: Yeah this could be tidied a bit. Most of the tar stuff here is from a libarchive example quickly modified.

**Comment by dtrudg** [comment]
@GodloveD thanks for the comments. Traveling light for Christmas and New Year in the UK with just an iPad right now, so will likely make changes after get back on 7th Jan.

File: src/docker-extract.c
Comment (GodloveD) [reply src/docker-extract.c:25]: OK.  As long as it is working as expected. Thanks for clarification. 😸

File: src/docker-extract.c
Comment (dtrudg) [reply src/docker-extract.c:212]: AFAIK the value here doesn't really matter, except for performance reasons.  In the IO callback docs they specify:

>To read an archive, you need to provide libarchive with a _read callback_ function. The read callback is expected to provide a _block_ of data to libarchive each time it is called.

>Blocks can be of any size or alignment at all. There are only two requirements:

>Each block must be at least one byte long. (A zero-byte block indicates end-of-file.)
When you return a pointer to a block, the memory where the block is stored must be stable until the next time your read function is called.

https://github.com/libarchive/libarchive/wiki/LibarchiveIO

From there it goes on to mention performance tuning, matching hardware block size etc. but there is no hard requirement on the block size.

For libarchive 3+ it may not use the specified value - but will inspect the archive for a block size to use:

>Note 1: Beginning with libarchive 3.0, the `archive_read_open_filename()` function inspects the file before deciding how to handle the blocksize. If the filename provided refers to a tape device, for example, it will use exactly the blocksize you specify. For other devices, it may adjust the requested blocksize in order to obtain better performance.

https://github.com/libarchive/libarchive/wiki/Examples

The 10240 is referenced elsewhere in docs as default for *write* for compatibility with old tar versions, and this is taken from the 'universal reader' example.

Comment (GodloveD) [reply src/docker-extract.c:212]: Makes sense.  Thanks for the clarification and the pointer to the docs.  😸

**Comment by dtrudg** [comment]
@GodloveD - I've had a chance to make some changes here, tidying it up. Let me know if there's anything else you'd like to see.

File: src/docker-extract.c
Comment (GodloveD) [inline src/docker-extract.c:191]: Actually I think that pathname should be declared at the beginning of the for loop where it is used since that is the only scope it is needed for.

**Review by GodloveD — COMMENTED** [review]
I'm mostly happy with this, although I think that variable declaration is still not happening in exactly the right place.  But I'm going to leave that for a more experienced C programmer like @yhcote to review.  If he or another of our C devs approves this I am happy to approve to and we can go ahead and get it merged. 

Also we are obviously going to need to make sure this is passing tests in Travis and that all conflicts are resolved before merging.  I'm not totally sure why I'm not seeing any passed or failed travis tests  on this 😕

**Comment by dtrudg** [comment]
@GodloveD - the travis test on Ubuntu seems to be passing now, but the CentOS builds are failing due to squashfs-tools deps. Not sure exactly how best to change this so it works but is consistent with the previous stuff you mentioned on changing approach to deps?

**Comment by GodloveD** [comment]
Hmm, yeah I noticed that.  squashfs-tools does need to be present on the host afaik.  Maybe adding it back in will lead to a passing build?

**Comment by dtrudg** [comment]
Well 2 green lights now, but CentOS7 failed with what I think might be a docker pull net issue and not related to the changes?

```
    raise BadStatusLine(line)
httplib.BadStatusLine: ''
Cleaning up...
Full output in: /tmp/stest.FSX6za
ERROR: Failed running test: 40-privblock.sh
```

I'll leave it to the eyes of the C and packaging gods now...

**Comment by jmstover** [comment]
Looks like a network, or remote server error. I've restarted the failed
CentOS7 job to see if it succeeds with a bit of time here between runs.

-J

**Comment by jmstover** [comment]
Wow... there's some uhh... issues:

https://travis-ci.org/singularityware/singularity/jobs/327450386

-J

**Comment by GodloveD** [comment]
OK.  Looks like MongoDB got it's act together and things are passing now.

File: src/docker-extract.c
Comment (yhcote) [inline src/docker-extract.c:36]: To simplify and because that code doesn't run multi-threaded, I would define a static buffer for string manipulation and I would make sure snprintf() is able to NULL terminate the string. Here's what you can do instead : 

int apply_opaque(const char *opq_marker, char *rootfs_dir) {                                                                 
    int retval;                                                                                                              
    char *token;                                                                                                             
    static char buf[MAXPATHLEN];                                                                                             
                                                                                                                             
    token = strrchr(opq_marker, '/');                                                                                        
    if (token == NULL) {                                                                                                     
        singularity_message(ERROR, "Error getting dirname for opaque marker\n");                                             
        ABORT(255);                                                                                                          
    }                                                                                                                        
                                                                                                                             
    retval = snprintf(buf, sizeof(buf), "%s/%s", rootfs_dir, opq_marker);                                                    
    if (retval == -1 || retval >= sizeof(buf)) {                                                                             
        singularity_message(ERROR, "Error with pathname too long\n");                                                        
        ABORT(255);                                                                                                          
    }                                                                                                                        
                                                                                                                             
    if (is_dir(buf) == 0)                                                                                                    
        s_rmdir(buf);                                                                                                        
                                                                                                                             
    return 0;                                                                                                                
}

File: src/docker-extract.c
Comment (yhcote) [inline src/docker-extract.c:55]: For this function also, manipulating strings with snprintf and using buffers, you can use the example I wrote to make it more simple.

**Review by yhcote — CHANGES_REQUESTED** [review]
The main concern is to make sure that a string is always null terminated with snprintf(). Restricting the copy length with the "size" param of snprintf() is not enough to guarantee that the resulting string will be null terminated if the string to be copied is larger than the buffer to contain it.

File: src/docker-extract.c
Comment (dtrudg) [reply src/docker-extract.c:36]: Thanks for the pointers - embarrassing how much sensible C I've forgotten living in Python etc. Just grepped through some C code from way back when I was 17 and found lots of `char buf[MAXLINE]` etc. Need to reactivate that part of my memory somehow 😨

File: src/docker-extract.c
Comment (dtrudg) [reply src/docker-extract.c:55]: Hopefully c8baf0d is something more like it for this too.

**Review by yhcote — APPROVED** [review]
Looks good to me

**Comment by dtrudg** [comment]
@GodloveD - do you need any more changes? Thanks.

File: .travis/rpmbuild_test
Comment (GodloveD) [inline .travis/rpmbuild_test:8]: Remind me, did we need squashfs-tools here for the build to complete?  If not, I think we should remove it.

Comment (GodloveD) [reply .travis/rpmbuild_test:8]: nvrmnd.  I see that we do need it.  Sorry for the noise.

Comment (vsoch) [reply .travis/rpmbuild_test:8]: +1, I've hit this bug a bunch of times forgetting to install it :)

Comment (dtrudg) [reply .travis/rpmbuild_test:8]: Yeah...

>GodloveD commented 8 days ago
Hmm, yeah I noticed that. squashfs-tools does need to be present on the host afaik. Maybe adding it back in will lead to a passing build?

>dctrud Add back squashfs-tools to rpmbuild_test script Verified 6ba0a1b

It was needed previously, had to add back in to make the build complete and get a travis checkmark. But... maybe this has been fixed by other things from the merges into development since then? Can't check right now if removing it is okay - but can tonight if needed.

**Review by GodloveD — APPROVED** [review]

File: .travis/rpmbuild_test
Comment (dtrudg) [reply .travis/rpmbuild_test:8]: Ahh okay sounds like its needed  :-)

Comment (GodloveD) [reply .travis/rpmbuild_test:8]: Things actually may have changed because @jmstover did some work recently to clean up dependencies in our test suite.  But don't worry I'm having a look myself.  Will let you know.  Thanks!

**Comment by GodloveD** [comment]
OK.  squashfs-tools is totally still needed. 
https://travis-ci.org/GodloveD/singularity/jobs/330564707
Thanks for your hard work and patience on this @dctrud!  And thanks for the useful comments and reviews @yarikoptic and @vsoch !

**Comment by olifre** [comment]
> OK. squashfs-tools is totally still needed.

Yeah. But installing it before building the package is not needed, so this was the wrong fix. 
As you can see in #1204 , the correct fix would have been to use `yum install -y` to install the built rpms, which automatically pulls in `squashfs-tools` instead of hardcoding the dependency and thus making the`rpmbuild` test useless. 
I'll rebase #1204 on current `development` tomorrow and revert this broken commit then as part of that PR.

**Comment by GodloveD** [comment]
Great.  Thanks @olifre.  Still trying to wrap my head around the subtleties of packaging.

**Pull request merged**
